web_advanced_filter/models/ir_filters.py

```diff
@@ -0,0 +1,215 @@
+                        eval_n(c.domain),
+                    ).ids
+                    domain = expression.AND([
+                        domain,
+                        [('id', 'not in', matching_ids)],
+                    ])
+                else:
+                    domain = expression.AND([
+                        domain,
+                        ['!'] + eval_n(c['domain']),
+                    ])
+            this.domain = str(expression.normalize_domain(domain))
+
+    @api.multi
+    def _inverse_domain(self):
+        for this in self:
+            this.domain_this = this.domain
+
+    @api.multi
+    @api.depends('domain')
+    def _compute_flags(self):
+        """check if this filter contains references to x2many fields. If so,
+        then negation goes wrong in nearly all cases, so we evaluate the
+        filter and remove its resulting ids"""
+        for this in self:
+            this.evaluate_before_negate = False
+            this.evaluate_always = False
+            domain = expression.normalize_domain(
+                safe_eval(this.domain or 'False') or [expression.FALSE_LEAF]
+            )
+            for arg in domain:
+                current_model = self.env.get(this.model_id)
+                if current_model is None:
+                    _logger.error(
+                        'Unknown model %s used in filter', this.model_id)
+                    continue
+                if not expression.is_leaf(arg) or not isinstance(arg[0], str):
+                    continue
+                has_x2many = False
+                has_auto_join = False
+                for field_name in arg[0].split('.'):
+                    if field_name in models.MAGIC_COLUMNS:
+                        continue
+                    field = current_model._fields[field_name]
+                    has_x2many |= field.type in self._evaluate_before_negate
+                    has_x2many |= bool(field.compute)
+                    has_auto_join |= getattr(field, 'auto_join', False)
+                    has_auto_join |= bool(field.compute)
+                    if field.comodel_name:
+                        current_model = self.env.get(field.comodel_name)
+                    if current_model is None or has_x2many and has_auto_join:
+                        break
+                this.evaluate_before_negate |= has_x2many
+                this.evaluate_always |= has_auto_join
+                if this.evaluate_before_negate and this.evaluate_always:
+                    break
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for values in vals_list:
+            values.setdefault(
+                'user_id',
+                False if values.get('save_as_public') else self.env.user.id,
+            )
+        return super(IrFilters, self).create(vals_list)
+
+    @api.multi
+    def _evaluate(self):
+        self.ensure_one()
+        return self.env[self.model_id].with_context(
+            safe_eval(self.context)
```

Comment:
This destroys the context of the caller, including the `lang` and `tz` keys. The default for `context` is `'{}'`.
Perhaps:
```suggestion
            **safe_eval(self.context)
```
I'm also wondering if the old context should be restored after the `.search()`.